api/server/router/system/system_routes.go

```diff
@@ -89,25 +89,60 @@ func (s *systemRouter) getVersion(ctx context.Context, w http.ResponseWriter, r
 	return httputils.WriteJSON(w, http.StatusOK, info)
 }
 
+// DiskUsageObject represents an object type used for disk usage query filtering.
+type DiskUsageObject string
+
+const (
```

Comment:
I just realised; should we move these consts out of the `api/server` and into `api/types` to make them available to clients (in case they want to use them)?

Did a quick move with that (not a fan of that package being just `types` (but we're missing a `system` package in `api/types` currently), but at least it makes it consumable by clients.

Diff / patch below

<details>

```patch
diff --git a/api/server/router/system/backend.go b/api/server/router/system/backend.go
index 2ddd5e0018..2d18d389da 100644
--- a/api/server/router/system/backend.go
+++ b/api/server/router/system/backend.go
@@ -13,7 +13,7 @@ import (
 // DiskUsageOptions holds parameters for system disk usage query.
 type DiskUsageOptions struct {
 	// ObjectTypes to include when collecting usage.
-	ObjectTypes map[DiskUsageObject]struct{}
+	ObjectTypes map[types.DiskUsageObject]struct{}
 }
 
 // Backend is the methods that need to be implemented to provide
diff --git a/api/server/router/system/system_routes.go b/api/server/router/system/system_routes.go
index b67fa2d870..e451e1ab7c 100644
--- a/api/server/router/system/system_routes.go
+++ b/api/server/router/system/system_routes.go
@@ -89,31 +89,17 @@ func (s *systemRouter) getVersion(ctx context.Context, w http.ResponseWriter, r
 	return httputils.WriteJSON(w, http.StatusOK, info)
 }
 
-// DiskUsageObject represents an object type used for disk usage query filtering.
-type DiskUsageObject string
-
-const (
-	// ContainerObject represents a container DiskUsageObject.
-	ContainerObject DiskUsageObject = "container"
-	// ImageObject represents an image DiskUsageObject.
-	ImageObject DiskUsageObject = "image"
-	// VolumeObject represents a volume DiskUsageObject.
-	VolumeObject DiskUsageObject = "volume"
-	// BuildCacheObject represents a build-cache DiskUsageObject.
-	BuildCacheObject DiskUsageObject = "build-cache"
-)
-
 func (s *systemRouter) getDiskUsage(ctx context.Context, w http.ResponseWriter, r *http.Request, vars map[string]string) error {
 	if err := httputils.ParseForm(r); err != nil {
 		return err
 	}
 
-	var objectTypes map[DiskUsageObject]struct{}
+	var objectTypes map[types.DiskUsageObject]struct{}
 	if ss, ok := r.Form["types"]; ok {
-		objectTypes = make(map[DiskUsageObject]struct{}, 4)
-		for _, s := range ss {
-			switch typ := DiskUsageObject(s); typ {
-			case ContainerObject, ImageObject, VolumeObject, BuildCacheObject:
+		objectTypes = make(map[types.DiskUsageObject]struct{}, 4)
+		for _, ts := range ss {
+			switch typ := types.DiskUsageObject(ts); typ {
+			case types.ContainerObject, types.ImageObject, types.VolumeObject, types.BuildCacheObject:
 				objectTypes[typ] = struct{}{}
 			default:
 				return invalidRequestError{fmt.Errorf("unknown object type: %s", typ)}
@@ -133,7 +119,7 @@ func (s *systemRouter) getDiskUsage(ctx context.Context, w http.ResponseWriter,
 	})
 
 	var buildCache []*types.BuildCache
-	if _, ok := objectTypes[BuildCacheObject]; ok || len(objectTypes) == 0 {
+	if _, ok := objectTypes[types.BuildCacheObject]; ok || len(objectTypes) == 0 {
 		eg.Go(func() error {
 			var err error
 			buildCache, err = s.builder.DiskUsage(ctx)
diff --git a/api/types/types.go b/api/types/types.go
index e3a159912e..2e2b02e62a 100644
--- a/api/types/types.go
+++ b/api/types/types.go
@@ -530,6 +530,20 @@ type ShimConfig struct {
 	Opts   interface{}
 }
 
+// DiskUsageObject represents an object type used for disk usage query filtering.
+type DiskUsageObject string
+
+const (
+	// ContainerObject represents a container DiskUsageObject.
+	ContainerObject DiskUsageObject = "container"
+	// ImageObject represents an image DiskUsageObject.
+	ImageObject DiskUsageObject = "image"
+	// VolumeObject represents a volume DiskUsageObject.
+	VolumeObject DiskUsageObject = "volume"
+	// BuildCacheObject represents a build-cache DiskUsageObject.
+	BuildCacheObject DiskUsageObject = "build-cache"
+)
+
 // DiskUsage contains response of Engine API:
 // GET "/system/df"
 type DiskUsage struct {
diff --git a/daemon/disk_usage.go b/daemon/disk_usage.go
index 9d345c905f..024fe7f3b5 100644
--- a/daemon/disk_usage.go
+++ b/daemon/disk_usage.go
@@ -23,20 +23,18 @@ func (daemon *Daemon) SystemDiskUsage(ctx context.Context, opts system.DiskUsage
 		images     []*types.ImageSummary
 		layersSize int64
 		volumes    []*types.Volume
-
-		err error
+		err        error
 	)
-	objectTypes := opts.ObjectTypes
-	if len(objectTypes) == 0 {
-		objectTypes = map[system.DiskUsageObject]struct{}{
-			system.ContainerObject: {},
-			system.ImageObject:     {},
-			system.VolumeObject:    {},
+	if len(opts.ObjectTypes) == 0 {
+		opts.ObjectTypes = map[types.DiskUsageObject]struct{}{
+			types.ContainerObject: {},
+			types.ImageObject:     {},
+			types.VolumeObject:    {},
 		}
 	}
-	for typ := range objectTypes {
+	for typ := range opts.ObjectTypes {
 		switch typ {
-		case system.ContainerObject:
+		case types.ContainerObject:
 			// Retrieve container list
 			containers, err = daemon.Containers(&types.ContainerListOptions{
 				Size: true,
@@ -46,7 +44,7 @@ func (daemon *Daemon) SystemDiskUsage(ctx context.Context, opts system.DiskUsage
 				return nil, fmt.Errorf("failed to retrieve container list: %v", err)
 			}
 
-		case system.ImageObject:
+		case types.ImageObject:
 			// Get all top images with extra attributes
 			images, err = daemon.imageService.Images(filters.NewArgs(), false, true)
 			if err != nil {
@@ -58,13 +56,13 @@ func (daemon *Daemon) SystemDiskUsage(ctx context.Context, opts system.DiskUsage
 				return nil, err
 			}
 
-		case system.VolumeObject:
+		case types.VolumeObject:
 			volumes, err = daemon.volumes.LocalVolumesSize(ctx)
 			if err != nil {
 				return nil, err
 			}
 
-		case system.BuildCacheObject:
+		case types.BuildCacheObject:
 			// NOTE: build-cache data usage is currently collected via
 			// direct call to buildkit.Builder.DiskUsage in systemRouter.getDiskUsage().
 
```

</details>